File: src/sdks/typescript/src/ports.ts
import { unwrap, type ApiClient } from './api'
import type { components } from './api/schema.gen'
import {
  RequestPipeline,
  type RequestDefaults,
  type RequestOptions,
} from './http'

/** One port's exposure record: visibility plus the port's public URL. */
export type PortExposure = components['schemas']['PortExposure']

/**
 * `'private'`: reachable with a capability token whose scope permits
 * the port — the default for every port. `'public'`: anonymous traffic
 * is admitted to this tenant application port. The server rejects its
 * reserved agent port.
 */
export type PortVisibility = components['schemas']['PortVisibility']

/**
 * Module for managing one sandbox's port exposure, available as
 * `sandbox.ports`.
 */
export class SandboxPorts {
  private readonly requests: RequestPipeline

  /**
   * Creates a port module bound to one sandbox.
   *
   * @param api Control-plane client.
   * @param sandboxId Sandbox identifier.
   * @param defaults Timeout, retry budget, and logger.
   */
  constructor(
    private readonly api: ApiClient,
    private readonly sandboxId: string,
    defaults?: RequestDefaults,
  ) {
    this.requests = new RequestPipeline(defaults)
  }

  /**
   * List the sandbox's exposure records.
   *
   * A port with no record is simply private; absence here does not mean
   * unreachable.
   *
   * @param options Per-call timeout, signal, and headers.
   * @returns The sandbox's exposure records.
   */
  async list(options?: RequestOptions): Promise<PortExposure[]> {
    const page = await this.requests.send(
      {
        operation: 'GET /v1/sandboxes/{sandboxId}/ports',
        idempotent: true,
        options,
      },
      async (init) =>
        unwrap(
          await this.api.GET('/v1/sandboxes/{sandboxId}/ports', {
            params: { path: { sandboxId: this.sandboxId } },
            ...init,
          }),
        ),
    )
    return page.items
  }

  /**
   * Record a port's exposure. Idempotent: repeating the call re-asserts
   * the record.
   *
   * @param port Port number inside the sandbox, from 1 through 65535.
   * @param visibility Access policy for the port.
   * @param options Per-call timeout, signal, and headers.
   * @returns The exposure record, including the port's public URL.
   *
   * @example
   * ```ts
   * const exposure = await sandbox.ports.expose(3000, 'public')
   * console.log(exposure.url) // https://3000-<sandboxId>.<domain>
   * ```
   */
  async expose(
    port: number,
    visibility: PortVisibility,
    options?: RequestOptions,
  ): Promise<PortExposure> {
    return this.requests.send(
      {
        operation: 'PUT /v1/sandboxes/{sandboxId}/ports/{port}',
        idempotent: true,
        options,
      },
      async (init) =>
        unwrap(
          await this.api.PUT('/v1/sandboxes/{sandboxId}/ports/{port}', {
            params: { path: { sandboxId: this.sandboxId, port } },
            body: { visibility },
            ...init,
          }),
        ),
    )
  }

  /**
   * Remove a port's exposure record, returning the port to the default:
   * private, capability token required. Public traffic to the port
   * stops.
   *
   * @param port Port number whose record to remove.
   * @param options Per-call timeout, signal, and headers.
   */
  async unexpose(port: number, options?: RequestOptions): Promise<void> {
    await this.requests.send(
      {
        operation: 'DELETE /v1/sandboxes/{sandboxId}/ports/{port}',
        idempotent: true,
        options,
      },
      async (init) =>
        unwrap(
          await this.api.DELETE('/v1/sandboxes/{sandboxId}/ports/{port}', {
            params: { path: { sandboxId: this.sandboxId, port } },
            ...init,
          }),
        ),
    )
  }
}

File: src/sdks/rust/src/ports.rs
//! Published-port exposure records.

use std::sync::Arc;

use reqwest::Method;

use crate::error::Error;
use crate::http::Http;
use crate::models::{ExposePortRequest, PortExposure, PortList, PortVisibility};

/// Port exposure records for one sandbox. Obtained from
/// [`crate::Sandbox::ports`].
///
/// Every port defaults to private (capability token required); public
/// exposure is an explicit, durable, audited opt-in. The server rejects
/// exposure records for its reserved agent port.
#[derive(Clone, Debug)]
pub struct Ports {
    http: Arc<Http>,
    sandbox_id: String,
}

impl Ports {
    pub(crate) fn new(http: Arc<Http>, sandbox_id: String) -> Self {
        Self { http, sandbox_id }
    }

    fn path(&self, suffix: &str) -> String {
        format!("/v1/sandboxes/{}/ports{suffix}", self.sandbox_id)
    }

    fn port_path(&self, port: u16) -> String {
        format!("/v1/sandboxes/{}/ports/{port}", self.sandbox_id)
    }

    /// The sandbox's exposure records. A port with no record is
    /// private.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the request is rejected or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn list(&self) -> Result<PortList, Error> {
        let request = self.http.request(Method::GET, &self.path(""));
        self.http.send_json(request).await
    }

    /// Record the port's exposure. Idempotent: repeating re-asserts
    /// the record.
    ///
    /// [`PortVisibility::Public`] admits anonymous traffic to this tenant
    /// application port; [`PortVisibility::Private`] records it without
    /// widening access.
    ///
    /// Returns the resulting exposure record.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the exposure is rejected or
    /// [`Error::Http`] if transport or decoding fails.
    pub async fn expose(
        &self,
        port: u16,
        visibility: PortVisibility,
    ) -> Result<PortExposure, Error> {
        let request = self
            .http
            .request(Method::PUT, &self.port_path(port))
            .json(&ExposePortRequest { visibility });
        self.http.send_json(request).await
    }

    /// Remove the port's exposure record, returning it to the
    /// default: private, capability token required. Public traffic to
    /// the port stops.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Api`] if the record cannot be removed or
    /// [`Error::Http`] if transport fails.
    pub async fn unexpose(&self, port: u16) -> Result<(), Error> {
        let request = self.http.request(Method::DELETE, &self.port_path(port));
        self.http.send_no_content(request).await
    }
}
